doc/dev/adr/0020-error-modeling-headers.md

```diff
@@ -0,0 +1,103 @@
+        };
+
+        // Specify error code, but no error payload  
+        response.WithApplicationError("negativeValue");
+
+        // Or you can specify error code and error payload
+        response.WithApplicationError(
+            "negativeValue",
+            new CustomErrorPayload() { ArgumentValue = request.IncrementValue },
+            new ErrorPayloadJsonSerializer(new TestEnvoys.Utf8JsonSerializer()));
+
+        return response;
+    }
+
+    // happy path omitted
+}
+```
+
+### Command invoker side
+
+```csharp
+public void main()
+{
+    await using MqttSessionClient mqttInvoker = await ClientFactory.CreateSessionClientFromEnvAsync();
+    await using CounterClient counterClient = new CounterClient(applicationContext, mqttInvoker);
+
+    IncrementRequestPayload payload = new IncrementRequestPayload
+    {
+        IncrementValue = -1
+    };
+
+    var response = await counterClient.Increment(executorId, payload).WithMetadata();
+    
+    // Check the RPC response for an application error
+    if (response.TryGetApplicationError(SomeJsonSerializer, out string? errorCode, out CustomErrorPayload? errorPayload))
+    {
+        // use the error code and strongly typed custom error payload type as wanted.
+    }
+
+    // Alternatively just fetch the error code if you don't want to deserialize the payload yet.
+    if (response.TryGetApplicationError(out string? errorCode))
+    {
+        // ...
+    }
+}
+```
+
+### Enforcement
+
+Note that, while this decision creates a standardized way of communicating application errors, our SDK will **not** enforce that users communicate application errors with this standard. Users will still be able to model application errors in response payloads or in custom user properties if they prefer. There is no way to force users to communicate errors with this standard pattern.
+
+## Samples
+
+We want to establish this pattern as the standard for communicating RPC application errors. To do so, we must either update or add a sample per-language that demonstrates handling at least one application error.
+
+For instance, we would make the counter service return an application error if the invocation specifies a negative value to increment by. This change won't include any DTDL level changes yet.
+
+## Open Questions
+
+- MQTT user properties are strings, not bytes, so is there an issue with using UTF8 encoding to convert a serialized payload to a string? Does this work for the serializers we already support (protobuf, avro, raw, etc.)
+  - Maybe this suggests the serialization interface we use here converts object <-> string rather than object <-> bytes
+
+- MQTT user property value size limit considerations? Users may provide very large payload objects and MQTT as a protocol may not be built for that?
+
+- Do we remove support for modeling application errors in the MQTT message payload as described in (ADR 19)[./0019-codegen-user-errs.md]? No one has onboarded to this model yet
```

Comment:
I believe @abhipsaMisra has begun using the new DTDL user-error modeling mechanisms in the Mqtt v3 extension.

Reply:
Good point. I do like that her use case allows for modeling of the error in the payload. The only question then becomes "do we care that there are 2 ways to model errors in DTDL?"

I know that we haven't really investigated modeling headers in DTDL land, but do you think it is feasible for us to support DTDL modeling of errors in either (and maybe both?) headers and MQTT payloads?

From the pure command invoker/executor side of things, it's certainly doable and not much overhead.

Reply:
It is certainly feasible from a technical perspective.  But having multiple ways to express error conditions runs counter to the goal (as I understand it) of ensuring commonality across applications.